<?xml version="1.0"?>
<launch>
    <!-- MAVROS posix SITL environment launch script -->
    <!-- launches Gazebo environment and 2x: MAVROS, PX4 SITL, and spawns vehicle -->
    <!-- vehicle model and world -->
    <arg name="est" default="ekf2"/>
    <!-- <arg name="vehicle_1" default="iris"/> -->
    <arg name="vehicle_1" default="typhoon_h480"/>
    <arg name="vehicle_2" default="solo"/>
    <arg name="world" default="$(find coverage)/worlds/simulation.world"/>
    <!-- gazebo configs -->
    <arg name="gui" default="true"/>
    <arg name="debug" default="false"/>
    <arg name="verbose" default="false"/>
    <arg name="paused" default="false"/>
    <!-- Gazebo sim -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="gui" value="$(arg gui)"/>
        <arg name="world_name" value="$(arg world)"/>
        <arg name="debug" value="$(arg debug)"/>
        <arg name="verbose" value="$(arg verbose)"/>
        <arg name="paused" value="$(arg paused)"/>
    </include>

    <arg name="z" default="0.5"/>

    <!-- UAV0 -->
    <group ns="uav0">
        <arg name="ID" value="0"/>
        <arg name="fcu_url" default="udp://:14540@localhost:14580"/>
        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="2"/>
            <arg name="y" value="2"/>
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="0.785398"/>
            <arg name="vehicle" value="$(arg vehicle_1)"/>
            <arg name="mavlink_udp_port" value="14560"/>
            <arg name="mavlink_tcp_port" value="4560"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <!-- UAV1 -->
    <group ns="uav1">
        <arg name="ID" value="1"/>
        <arg name="fcu_url" default="udp://:14541@localhost:14581"/>
        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="23"/>
            <arg name="y" value="2"/>
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="2.35619"/>
            <arg name="vehicle" value="$(arg vehicle_1)"/>
            <arg name="mavlink_udp_port" value="14561"/>
            <arg name="mavlink_tcp_port" value="4561"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <!-- UAV2 -->
    <group ns="uav2">
        <arg name="ID" value="2"/>
        <arg name="fcu_url" default="udp://:14542@localhost:14582"/>
        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="2"/>
            <arg name="y" value="23"/>
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="-0.785398"/>
            <arg name="vehicle" value="$(arg vehicle_1)"/>
            <arg name="mavlink_udp_port" value="14562"/>
            <arg name="mavlink_tcp_port" value="4562"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <!-- UAV3 -->
    <group ns="uav3">
        <arg name="ID" value="3"/>
        <arg name="fcu_url" default="udp://:14543@localhost:14583"/>
        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="23"/>
            <arg name="y" value="23"/>
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="3.92699"/>
            <arg name="vehicle" value="$(arg vehicle_1)"/>
            <arg name="mavlink_udp_port" value="14563"/>
            <arg name="mavlink_tcp_port" value="4563"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <!-- UAV4 -->
<!--     <group ns="uav4">
        
        <arg name="ID" value="4"/>
        <arg name="fcu_url" default="udp://:14544@localhost:14584"/>

        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="12.5"/>
            <arg name="y" value="2.0"/>
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="1.5713"/>
            <arg name="vehicle" value="$(arg vehicle_1)"/>
            <arg name="mavlink_udp_port" value="14564"/>
            <arg name="mavlink_tcp_port" value="4564"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>

        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group> -->

    <!-- UAV10 -->
    <group ns="uav4">

        <arg name="ID" value="4"/>
        <arg name="fcu_url" default="udp://:14544@localhost:14584"/>

        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="5.0"/>
            <arg name="y" value="14.0"/>
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="0"/>
            <arg name="vehicle" value="$(arg vehicle_2)"/>
            <arg name="mavlink_udp_port" value="14564"/>
            <arg name="mavlink_tcp_port" value="4564"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>

        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>


    <!-- UAV5 -->
    <group ns="uav5">

        <arg name="ID" value="5"/>
        <arg name="fcu_url" default="udp://:14545@localhost:14585"/>

        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="6.0"/>
            <arg name="y" value="6.0"/>
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="0"/>
            <arg name="vehicle" value="$(arg vehicle_2)"/>
            <arg name="mavlink_udp_port" value="14565"/>
            <arg name="mavlink_tcp_port" value="4565"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>

        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <!-- UAV6 -->
    <group ns="uav6">

        <arg name="ID" value="6"/>
        <arg name="fcu_url" default="udp://:14546@localhost:14586"/>

        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="7.0"/>
            <arg name="y" value="7.0"/>
<!--             <arg name="x" value="18.0"/>
            <arg name="y" value="8.0"/> -->
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="0"/>
            <arg name="vehicle" value="$(arg vehicle_2)"/>
            <arg name="mavlink_udp_port" value="14566"/>
            <arg name="mavlink_tcp_port" value="4566"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>

        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <!-- UAV7 -->
    <group ns="uav7">

        <arg name="ID" value="7"/>
        <arg name="fcu_url" default="udp://:14547@localhost:14587"/>

        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="14.0"/>
            <arg name="y" value="11.5"/>
<!--             <arg name="x" value="13.0"/>
            <arg name="y" value="13.0"/> -->
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="0"/>
            <arg name="vehicle" value="$(arg vehicle_2)"/>
            <arg name="mavlink_udp_port" value="14567"/>
            <arg name="mavlink_tcp_port" value="4567"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>

        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <!-- UAV8 -->
    <group ns="uav8">

        <arg name="ID" value="8"/>
        <arg name="fcu_url" default="udp://:14548@localhost:14588"/>

        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="14.0"/>
            <arg name="y" value="13.5"/>
            <!-- <arg name="x" value="15.0"/> -->
            <!-- <arg name="y" value="15.0"/> -->
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="0"/>
            <arg name="vehicle" value="$(arg vehicle_2)"/>
            <arg name="mavlink_udp_port" value="14568"/>
            <arg name="mavlink_tcp_port" value="4568"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>

        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <!-- UAV9 -->
    <group ns="uav9">

        <arg name="ID" value="9"/>
        <arg name="fcu_url" default="udp://:14549@localhost:14589"/>

        <include file="$(find px4)/launch/single_vehicle_spawn_sdf.launch">
            <arg name="x" value="15.3"/>
            <arg name="y" value="12.5"/>
            <!-- <arg name="x" value="19.0"/> -->
            <!-- <arg name="y" value="12.5"/> -->
            <arg name="z" value="$(arg z)"/>
            <arg name="R" value="0"/>
            <arg name="P" value="0"/>
            <arg name="Y" value="0"/>
            <arg name="vehicle" value="$(arg vehicle_2)"/>
            <arg name="mavlink_udp_port" value="14569"/>
            <arg name="mavlink_tcp_port" value="4569"/>
            <arg name="ID" value="$(arg ID)"/>
        </include>

        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
        </include>
    </group>

    <node name="ArmandOffboard" pkg="coverage" type="px4_mavros.py" output="screen" />
    <param name="uav0/vel_control" type="int" value="0" />
    <param name="uav1/vel_control" type="int" value="0" />
    <param name="uav2/vel_control" type="int" value="0" />
    <param name="uav3/vel_control" type="int" value="0" />
    <param name="uav4/vel_control" type="int" value="0" />
    <param name="uav5/vel_control" type="int" value="0" />
    <param name="uav6/vel_control" type="int" value="0" />
    <param name="uav7/vel_control" type="int" value="0" />
    <param name="uav8/vel_control" type="int" value="0" />
    <param name="uav9/vel_control" type="int" value="0" />

</launch>